Beginning OOP with Ruby

z64 opened the pull request:
I wrote this a long time ago, currently living in my Gist here: https://gist.github.com/z64/343d8c1dfcac593f1bcfbb7b6f63d962

Todo:
- [x] Make links actual links.
- [x] ~~Our inline code blocks kind of look like shit. Maybe improve this?~~ (Thanks @unleashy !)
- [x] Remove Gist post with notice that it's been moved to the blog

z64 committed:
Beginning OOP with Ruby

z64 committed:
fix link to discord

z64 committed:
format links into actual links

unleashy commented on _posts/2017-07-29-ruby-oop.md:
```diff
@@ -0,0 +1,381 @@
+
+The first thing we want to do to tackle this problem is to break apart the problem into a series of smaller, bite sized questions. "Good Ruby code" will read a lot like english - I think you'll find the syntax we end up writing will almost look like english answers to these questions. You probably already know the answers to these (which is why I picked this as an example), but read them and pay  ...
+
+**1) What is perimeter?**
```
Maybe this should be "what is **a** perimeter"?

z64 replied:
No, this is referring to perimeter as a concept, not a particular perimeter.

unleashy commented on _posts/2017-07-29-ruby-oop.md:
```diff
@@ -0,0 +1,381 @@
+
+## Breaking it down
+
+The first thing we want to do to tackle this problem is to break apart the problem into a series of smaller, bite sized questions. "Good Ruby code" will read a lot like english - I think you'll find the syntax we end up writing will almost look like english answers to these questions. You probably already know the answers to these (which is why I picked this as an example), but read them and pay  ...
```
Capitalize "english" you fuck

unleashy commented on _posts/2017-07-29-ruby-oop.md:
```diff
@@ -0,0 +1,381 @@
+
+**4) Okay, so what is a point?**
+
+> A point is a location in space. For our purposes, a location on a 2D plane, bound by **coordinates**.
```
I think "For our purposes, this location is on a 2D plane, bounded by **coordinates**." sounds better

unleashy commented on _posts/2017-07-29-ruby-oop.md:
````diff
@@ -0,0 +1,381 @@
+```ruby
+Polygon.new(___).perimeter
+```
+> In Ruby, and in most OOP languages, a `class` describes a "template" for an object. An object is usually a collection of data, and a set of methods or functions that allow you to interact with that data or give it logic. We first describe the objects with a `class`, and create new instances of them with `Thing.new`.
````
`Thing.new` feels broad, maybe say like "`X.new`, where X is the name of your class"? Not sure

z64 replied:
It's entirely meant to be broad. But in any case, I wouldn't chose `X` as the name might confuse with the concepts being presented in the tutorial. (X, Y coordinates)

unleashy commented on _posts/2017-07-29-ruby-oop.md:
````diff
@@ -0,0 +1,381 @@
+puts point.inspect #=> #<Point:0x2ae37f0 @x=1.5, @y=-3.0>
+```
+
+We're going to need to use these attributes of a point in order to do our calculations. To do so, we can define methods that will return the values of our attributes:
````
I think you need to explain why instance variables can't be accessed from outside, and why these methods are necessary. dunno

z64 replied:
Tricky without getting too off topic. I'll try.

unleashy commented on _posts/2017-07-29-ruby-oop.md:
```diff
@@ -0,0 +1,381 @@
+
+What else would be helpful to do with our point, or perhaps any other point? Well, with any two points, you can calculate the distance between them.
+
+Ruby has a rich standard library, of which includes mathematical functions. Needing the pythagorean theorem here, if we do some searching we can see Ruby already provides a method for us to calculate this instead of having to implement it ourselves:
```
"Needing the pythagorean theorem here, if we do some searching we can see ..."
Kinda tangled my brain a little. Something like "We need the Pythagorean theorem to calculate the distance between two points, and doing some searching we can see ..." would work better

unleashy commented on _posts/2017-07-29-ruby-oop.md:
````diff
@@ -0,0 +1,381 @@
+  end
+end
+```
+> Here, the `x` in `x_difference = x - other.x` is actually calling this objects *own* `attr_reader` method internally that fetches its `@x` attribute. In another language, you may write something like `self.x` to refer to this value. You can do that here too in Ruby, but in most cases you can omit `self.`, as it is already understood what you mean.
````
"..., but in most cases you can omit `self.`, as it is already understood what you mean."
Something like "but in most cases you can omit `self.`, because Ruby understands what you mean" if that doesn't sound cringy as fuck to you

unleashy commented on _posts/2017-07-29-ruby-oop.md:
````diff
@@ -0,0 +1,381 @@
+```
+> Arguments for methods can be given a default value using the above syntax in `def initialize`. Here, I'm making `segments` default to an empty array.
+
+We're going to design `Polygon` to accept an `Array` of `Line`.
````
You need to explain what an array is, maybe? /shrug

z64 replied:
I think that may be out of the scope of the tutorial (basic, high level OOP design) but I can add a small note. (i.e., this isn't a *beginner's Ruby* tutorial, it's an OOP tutorial)

unleashy replied:
O fuck right. I thought this was a beginner tutorial thing. aight

unleashy commented on _posts/2017-07-29-ruby-oop.md:
```diff
@@ -0,0 +1,381 @@
+
+We're going to design `Polygon` to accept an `Array` of `Line`.
+
+Our new `Polygon` can be instanced like so..
```
Instantiated *

unleashy reviewed (changes requested):
lambasted you

z64 committed:
fix English

z64 committed:
clarify classficiation of a point

z64 committed:
clarify pythag theorem

z64 committed:
clarify self. omission

z64 committed:
instantiated

z64 committed:
clarify attributes

z64 commented:
@Daniel-Worrall I plan to have this published by Friday. I'll accept edits after publishing, but it would be preferred to incorporate any of your suggestions beforehand if you can get around to it by then :+1: 

![image](https://user-images.githubusercontent.com/6119032/28894044-8c7ee940-77a1-11e7-9416-376958daaf91.png)

z64 committed:
add tags

unleashy commented:
R U B Y

T A C O

Daniel-Worrall commented on _posts/2017-07-29-ruby-oop.md:
```diff
@@ -0,0 +1,383 @@
+
+This post attempts to guide a beginner through object-oriented progamming in Ruby, showing them the thought process behind both designing and interacting with a library, and problem solving in general with Ruby.
+
+To accomplish this, we will be solving a simple geometry problem:
```
Probably reword this to "we will solve" to sound more like the intention of the post, rather than a plan to solve.

Daniel-Worrall replied:
Probably reword this to "we will solve" to sound more like the intention of the post, rather than a plan to solve.

Daniel-Worrall commented on _posts/2017-07-29-ruby-oop.md:
```diff
@@ -0,0 +1,383 @@
+
+> Perimeter is the total distance around an object, in our case a 2D polygon with any number of sides.
+
+**2) What is a Polygon?**
```
for all 4 of these points, we should keep consistent about capitalising the subjects. Either lowercase as speaking about them in general cases, or uppercase as preparation to refer to them as a class object. Also, we can use inline codeblocks if that is the intention.

Daniel-Worrall commented on _posts/2017-07-29-ruby-oop.md:
```diff
@@ -0,0 +1,383 @@
+
+> A point is a location in space. For our purposes, this location is on a 2D plane, bound by **coordinates**.
+
+In the next section, we'll convert these concepts into Ruby code. Please re-read this section once we're done and see how what we said here is directly represented in the answer we come up with.
```
Remove the please to make it more direct, and possibly add a part at the bottom that says to go back up and read this part. Maybe with a link too. Or move this second part to the end of the post.

Daniel-Worrall commented on _posts/2017-07-29-ruby-oop.md:
````diff
@@ -0,0 +1,383 @@
+Consider any kind of object around you. A pen, a car, your computer - they all have properties, or *attributes*.  Ruby objects can be given *attributes* that let you describe and store data inside them. We do this with the `@var` syntax, like so, describing what defines a `Point` in 2D space:
+
+```ruby
+# A location in 2D space.
````
Possibly call this position instead of location

z64 replied:
I refer to it as a location in the previous section. I think this semantic is fine.

Daniel-Worrall commented on _posts/2017-07-29-ruby-oop.md:
````diff
@@ -0,0 +1,383 @@
+puts point.x #=> 1.5
+puts point.y #=> -3.0
+```
+> You can read more about attributes, readers, setters, and more [here](http://ruby-for-beginners.rubymonstas.org/writing_classes/instance_variables.html).
````
Perhaps move this up between where you described what they are and when you use them.

z64 replied:
I think this is an appropriate place for this footer, as linking it earlier may distract someone from what I'm explaining if they open a link to the other site. Here, I've fully demonstrated attributes and getters/setters, and am about to move on to another topic ; this is a safe place to deviate from the lesson if someone wants to know more about them.

Daniel-Worrall commented on _posts/2017-07-29-ruby-oop.md:
````diff
@@ -0,0 +1,383 @@
+```
+> You can read more about attributes, readers, setters, and more [here](http://ruby-for-beginners.rubymonstas.org/writing_classes/instance_variables.html).
+
+What else would be helpful to do with our point, or perhaps any other point? Well, with any two points, you can calculate the distance between them.
````
I like how it's posed as a question, but this comes across very wordy. Could change to "What could we add to our `Point`? With two points, we can calculate the distance between them."

Daniel-Worrall commented on _posts/2017-07-29-ruby-oop.md:
````diff
@@ -0,0 +1,383 @@
+  end
+end
+```
+> Here, the `x` in `x_difference = x - other.x` is actually calling this objects *own* `attr_reader` method internally that fetches its `@x` attribute. In another language, you may write something like `self.x` to refer to this value. You can do that here too in Ruby, but in most cases you can omit `self.`, because Ruby understands what you mean.
````
remove "actually", then change second case of "Ruby" to "it" or similar. Could give an example of another language to give the user a reference to what other languages you could be referring to if Ruby is their first language.

z64 replied:
By saying "actually" here I'm attempting to distinguish it as a method call, with the assumption the reader might think it's just another temporary (local) variable.

Beyond this I'd rather not dive into the tangent of how other language's method / function scopes work.

Daniel-Worrall commented on _posts/2017-07-29-ruby-oop.md:
````diff
@@ -0,0 +1,383 @@
+end
+```
+
+From what we know, we can now give our `Line`s two attributes. We'll call them `start` and `stop` (I would use the word `end` here, but this is a Ruby keyword and it could get confusing.)
````
Maybe add to say that you cannot use Ruby keywords as variable names, not that it's confusing to use. (Yes, there are symbol workarounds, but I think it's better to talk about this rather than leave it vague)

Daniel-Worrall commented on _posts/2017-07-29-ruby-oop.md:
```diff
@@ -0,0 +1,383 @@
+# A finite segment (or distance) between two points.
+class Line
+  def length
+    start.distance stop
```
Explain how this line relates to the earlier code and that start and stop are points

Daniel-Worrall commented on _posts/2017-07-29-ruby-oop.md:
```diff
@@ -0,0 +1,383 @@
+
+Learn to read documentation like those on [http://www.rubydoc.info/](http://awesome-ruby.com/). Then you'll be a real library ninja, and can answer a lot of your own questions! (Hopefully!)
+
+When you're stuck, there's a lot of helpful ruby people out there. StackOverflow, IRC, etc. - or even me! I can be reached at wherever you're viewing this post, or at [my email](mailto:zachnowicki@gmail.com), or on [Discord](https://discordapp.com) - I'm `z64#2639`.
```
Could also add a bit about the blog.

Daniel-Worrall reviewed (approved):
Overall, decent write up. Just some style / language change suggestions. Will set this to approve as none of them are critical, but should at least be looked over.

z64 committed:
updates from feedback

clarify opening statement

make tone and capitilization consistent

add re-reading link, stronger wording

make segway into hypot stuff less wordy

clarify restriction on keywords as variable names

clarify start / stop usage in Line#length

add link to the rest of the blog
